src/AutoRest.CSharp/Common/Output/PostProcessing/PostProcessor.cs

```diff
@@ -0,0 +1,259 @@
+
+    protected virtual async Task<ModelSymbols> GetModelSymbolsAsync(Compilation compilation, bool publicOnly = true)
+    {
+        var result = new HashSet<INamedTypeSymbol>(SymbolEqualityComparer.Default);
+        var declarationCache = new Dictionary<INamedTypeSymbol, HashSet<BaseTypeDeclarationSyntax>>(SymbolEqualityComparer.Default);
+        var documentCache = new Dictionary<Document, HashSet<INamedTypeSymbol>>();
+        foreach (var document in _project.Documents)
+        {
+            if (!GeneratedCodeWorkspace.IsSharedDocument(document))
+            {
+                var root = await document.GetSyntaxRootAsync();
+                if (root == null)
+                    continue;
+
+                var semanticModel = compilation.GetSemanticModel(root.SyntaxTree);
+
+                foreach (var typeDeclaration in root.DescendantNodes().OfType<BaseTypeDeclarationSyntax>())
+                {
+                    var symbol = semanticModel.GetDeclaredSymbol(typeDeclaration);
+                    if (symbol == null)
+                        continue;
+                    if (publicOnly && symbol.DeclaredAccessibility != Accessibility.Public)
+                        continue;
+                    result.Add(symbol);
+                    declarationCache.AddInList(symbol, typeDeclaration);
+                    documentCache.AddInList(document, symbol, () => new HashSet<INamedTypeSymbol>(SymbolEqualityComparer.Default));
+                }
+            }
+        }
+
+        return new ModelSymbols(result, declarationCache, documentCache);
+    }
+
+    public async Task<Project> InternalizeAsync()
+    {
+        var compilation = await _project.GetCompilationAsync();
+        if (compilation == null)
+            return _project;
+
+        // first get all the declared symbols
+        var definitions = await GetModelSymbolsAsync(compilation, true);
+        // build the reference map
+        var referenceMap = await new ReferenceMapBuilder(compilation, _project, HasDiscriminator).BuildPublicReferenceMapAsync(definitions.DeclaredSymbols, definitions.DeclaredNodesCache);
+        // get the root symbols
+        var rootSymbols = GetRootSymbols(definitions);
+        // traverse all the root and recursively add all the things we met
+        var publicSymbols = TraverseModelsAsync(rootSymbols, referenceMap);
+
+        var symbolsToInternalize = definitions.DeclaredSymbols.Except(publicSymbols);
+
+        var nodesToInternalize = new List<BaseTypeDeclarationSyntax>();
+        foreach (var symbol in symbolsToInternalize)
+        {
+            nodesToInternalize.AddRange(definitions.DeclaredNodesCache[symbol]);
+        }
+
+        foreach (var model in nodesToInternalize)
+        {
+            MarkInternal(model);
+        }
+
+        return _project;
```

Comment:
As this class holds a ref of project , do we need to return project in InternalizeAsync() and RemoveAsync()?
It would make the calling place of those 2 APIs a bit strange.

Reply:
it is quite a common pattern that we return project whenever it gets changed because the project is immutable, we have to manually maintain its reference.
It is OK to make these two methods no longer return project and expose another property for `project`, but I do not like it because these two methods are changing it, and we need it to be reflected on the project instance right away.
For instance, if you call Internalize and you would like to change the project by yourself and then call remove, you will have to use `postProcessor.Project` first to get the reference. On the other hand, if you are not changing the project in the middle, just ignore the return value of it, it harms nothing.